Speedup BytesArray.indexOf

original-brownbear opened the pull request:
This method is only relevant for searching the line breaks in bulk requests. We can assume that in most cases the number of bytes between two line breaks is well above 8. This makes the SIMD(ish)/SWAR logic this PR introduces much faster than the simple loop.
Benchmarking this using the TSDB track indexing step, shows about a 3x to 4x speedup when searching for the next byte during bulk request parsing.

Profiling from running this on AWS ARM instances shows the massive relative reduction in cost from using the SWAR logic. The logic here is essentially the same as in Netty but optimized for searching `byte[]` directly was well and does not contain a branch for big endian.

Before:

<img width="2039" alt="image" src="https://github.com/elastic/elasticsearch/assets/6490959/09b698e0-b570-4d69-be05-4620e727e1a5">


After:
<img width="1371" alt="image" src="https://github.com/elastic/elasticsearch/assets/6490959/c198cff0-768a-4a90-a5d3-6bb1de654c60">

original-brownbear committed:
Speedup BytesArray.indexOf

This method is only relevant for searching the line breaks in bulk
requests. We can assume that in most cases the number of bytes between
two line breaks is well above 8. This makes the SIMD(ish) logic this PR
introduces much faster than the simple loop.
Benchmarking this using the TSDB track indexing step, shows about a 3x
to 4x speedup when searching for the next byte during bulk request
parsing.

elasticsearchmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

DaveCTurner reviewed (commented):
Seems fine although a few comments on how it works would be appreciated.

Also `org.elasticsearch.common.bytes.AbstractBytesReferenceTestCase#testIndexOf` seems a little weak - after the first match it only picks `from` as `positions.get(i-1) + 1`, and it doesn't check that we get a `-1` if `from` is after the last position. Could you strengthen this to improve coverage?

DaveCTurner commented:
> how it works

I particularly mean the bitbanging stuff in `findInLong`, this is the only bit that I can't do in my head, although a few more comments throughout would be helpful if only to avoid this stuff getting "simplified" away at a future date. Also SWAR is not a particularly searchable acronym, maybe spell it out and/or add some references.

DaveCTurner commented:
Sorry one more testing nit, `org.elasticsearch.common.bytes.BytesArrayTests#newBytesReference(int, int)` creates a slice with some leading junk but no trailing junk, the slice always ends at the end of the array. Let's have junk both sides sometimes please.

original-brownbear committed:
comments + tests

original-brownbear commented:
>  it doesn't check that we get a -1 if from is after the last position. Could you strengthen this to improve coverage?

Done, funny enough in the composite bytes reference we explicitly made this throw. I adjusted the code there now to return `-1` and added a check in `BytesArray`.

All other points should be addressed now as well :)

DaveCTurner commented on test/framework/src/main/java/org/elasticsearch/common/bytes/AbstractBytesReferenceTestCase.java:
```diff
@@ -667,6 +667,7 @@ public void testIndexOf() throws IOException {
         });
         final byte missing = randomValueOtherThanMany(map::containsKey, ESTestCase::randomByte);
         assertEquals(-1, bytesReference.indexOf(missing, randomIntBetween(0, Math.max(0, size - 1))));
+        assertEquals(-1, bytesReference.indexOf(missing, bytesReference.length() + randomIntBetween(0, 100)));
```
Ah no sorry clearly it's a bug to pass `from ≥ length`, that wasn't what I meant. It's coverage of other stuff with `from < length` I think we're missing:

```diff
diff --git a/test/framework/src/main/java/org/elasticsearch/common/bytes/AbstractBytesReferenceTestCase.java b/test/framework/src/main/java/org/elasticsearch/common/bytes/AbstractBytesReferenceTestCase.java
index d0004b76601..d2d2837ee7a 100644
--- a/test/framework/src/main/java/org/elasticsearch/common/bytes/AbstractBytesReferenceTestCase.java
+++ b/test/framework/src/main/java/org/elasticsearch/common/bytes/AbstractBytesReferenceTestCase.java
@@ -661,9 +661,13 @@ public abstract class AbstractBytesReferenceTestCase extends ESTestCase {
         map.forEach((value, positions) -> {
             for (int i = 0; i < positions.size(); i++) {
                 final int pos = positions.get(i);
-                final int from = i == 0 ? randomIntBetween(0, pos) : positions.get(i - 1) + 1;
+                final int from = randomIntBetween(i == 0 ? 0 : positions.get(i - 1) + 1, pos);
                 assertEquals(bytesReference.indexOf(value, from), pos);
             }
+            final int firstNotFoundPos = positions.get(positions.size() - 1) + 1;
+            if (firstNotFoundPos < bytesReference.length()) {
+                assertEquals(-1, bytesReference.indexOf(value, between(firstNotFoundPos, bytesReference.length() - 1)));
+            }
         });
         final byte missing = randomValueOtherThanMany(map::containsKey, ESTestCase::randomByte);
         assertEquals(-1, bytesReference.indexOf(missing, randomIntBetween(0, Math.max(0, size - 1))));
```

original-brownbear replied:
Right :) that makes more sense, adjusted the test and revert changes :)

original-brownbear committed:
Merge remote-tracking branch 'elastic/main' into much-faster-indexof

original-brownbear committed:
test fixed

DaveCTurner reviewed (approved):
LGTM

original-brownbear commented:
Thanks David!

franz1981 commented:
Well done @original-brownbear and happy that the SWAR method is become more used elsewhere:) (FYI, I'm the author of the PR in Netty to introduce it)

It could be worthy to cache and expose method APIs for common known delimiters too (saving the multiplication) or allow searching more than one in parallel - which is what StringZilla does

DaveCTurner commented on server/src/main/java/org/elasticsearch/common/bytes/BytesArray.java:
```diff
@@ -58,10 +58,87 @@ public byte get(int index) {
     @Override
     public int indexOf(byte marker, int from) {
-        for (int i = offset + from; i < offset + length; i++) {
-            if (bytes[i] == marker) {
```
Thanks @franz1981; @original-brownbear would you look into that?

original-brownbear replied:
Thanks as well @franz1981 will look into that :) And indeed ... we're only ever searching for the carriage return here I believe ... lets see what those two changes combined can get us :)
